lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp

```diff
@@ -431,6 +434,93 @@ bool circt::firrtl::fromJSON(json::Value &value, StringRef circuitTarget,
   return true;
 }
 
+bool circt::firrtl::fromOMIRJSON(json::Value &value, StringRef circuitTarget,
+                                 llvm::StringMap<ArrayAttr> &annotationMap,
+                                 json::Path path, CircuitOp circuit) {
+  auto *context = circuit.getContext();
+
+  // The JSON value must be an array of objects.  Anything else is reported as
+  // invalid.
+  auto *array = value.getAsArray();
+  if (!array) {
+    path.report(
+        "Expected OMIR to be an array of nodes, but found something else.");
+    return false;
+  }
+
+  // Build a mutable map of Target to Annotation.
+  SmallVector<Attribute> omnodes;
+  for (size_t i = 0, e = (*array).size(); i != e; ++i) {
+    auto *object = (*array)[i].getAsObject();
+    auto p = path.index(i);
+    if (!object) {
+      p.report("Expected OMIR to be an array of objects, but found an array of "
+               "something else.");
+      return false;
+    }
+
+    // Manually built up OMNode.
+    NamedAttrList omnode;
+
+    // Validate that this looks like an OMNode.  This should have three fields:
+    //   - "info": String
+    //   - "id": String that starts with "OMID:"
+    //   - "fields": Array<Object>
+    // Fields is optional and can be either an array or not there.  If not
+    // there, then construct an empty array for it.
+    auto maybeInfo = object->getString("info");
+    if (!maybeInfo) {
+      p.report("OMNode missing mandatory member \"info\" with type \"string\"");
+      return false;
+    }
+    auto maybeID = object->getString("id");
+    if (!maybeID || !maybeID.getValue().startswith("OMID:")) {
+      p.report("OMNode missing mandatory member \"id\" with type \"string\" "
+               "that starts with \"OMID:\"");
+      return false;
+    }
+    auto *maybeFields = object->get("fields");
+    if (maybeFields && !maybeFields->getAsArray()) {
+      p.report("OMNode has \"fields\" member with incorrect type (expected "
+               "\"array\")");
+      return false;
+    }
+    Attribute fields;
+    if (!maybeFields)
+      fields = ArrayAttr::get(context, {});
+    else
+      fields = convertJSONToAttribute(context, *maybeFields, p.field("fields"));
```

Comment:
Thinking about this some more I'm wondering whether it would make sense to turn this into a `DictionaryAttr` so we have an easier time looking up fields by name?

Reply:
Ah, great observation. This is really what this is trying to model (and internal code will cache the `name -> OMField` mapping for faster lookup later. I don't see why we can't just jump to a `DictionaryAttr` here.

Reply:
I got confirmation that while this is actually an array, it is modeling dictionary behavior. All `name` fields will be unique in `*.omir.json`. I'll apply this update in a rebase.